### PR TITLE
Creative API: resize - attempt 2

### PR DESCRIPTION
**Pull request by nekodo**
Adds support for requesting resize using embed-size messages from nested frames. This time should dodge the issues my previous patch had by no longer having the weird "fail the pending request" logic since attemptChangeSize is now more reasonable and takes a failure callback.

### CODE REVIEW

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 1]
```diff
@@ -148,29 +151,39 @@ export class AmpAdApiHandler {
...
-      postMessage(
-          this.iframe_,
-          'embed-size-denied',
-          {requestedHeight: height, requestedWidth: width},
-          targetOrigin,
-          this.is3p_);
-    });
+  updateSize_(height, width, source, origin) {
+    this.baseInstance_.attemptChangeSize(height, width,
+        () => this.sendEmbedSizeResponse_(
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: `#attemptChangeSize` doesn't take a success and failure callback, it returns a Promise. You'll have to put these in a `#then`:

```js
attemptChangeSize(height, width).then(success, failure);

**Comment by jridgewell** [comment]
Re: https://github.com/ampproject/amphtml/pull/4023

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 1 above)
Comment (sriramkrish85) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: +1

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 2]
```diff
@@ -148,29 +151,39 @@ export class AmpAdApiHandler {
...
+   * @param {number} requestedHeight
+   * @param {!Window} source
+   * @param {string} origin
+   * @private
+   */
+  sendEmbedSizeResponse_(
+      success, requestedWidth, requestedHeight, source, origin) {
+    postMessageToWindows(
+        this.iframe_,
+        [{win: source, origin}],
```
Comment (sriramkrish85) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: origin & source should already be present in the event object right? why do we have to pass them around?

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 1 above)
Comment (nekodo) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: You seem to be correct, but that is not what the code at head is doing: https://github.com/ampproject/amphtml/blob/master/extensions/amp-ad/0.1/amp-ad-api-handler.js#L159.

Is the code at head actually broken?

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 2 above)
Comment (nekodo) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: We need to tell postMessageToWindows which windows to send the message to. With this change we are no longer messaging only the contentWindow of the iframe, but potentially also any of the child windows within that iframe. That's the point of the change.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 1 above)
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: > Is the code at head actually broken?

😭. @zhouyx

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 2 above)
Comment (sriramkrish85) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: forgive my ignorance. Why do other windows care about resize response? only the requestor should benefit from it right?

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 1 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: 😢 My bad. messed up on rebase, #4258

**Comment by lannka** [comment]
LGTM. Defer to @sriramkrish85 for the final approval.

**Comment by jridgewell** [comment]
This'll need to be rebased after https://github.com/ampproject/amphtml/pull/4258.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 2 above)
Comment (sriramkrish85) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: CC @dvoytenko

**Comment by jridgewell** [comment]
Ping @lannka.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 2 above)
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: I think this is for a nested child iframe to send message to its ancestor.

**Comment by lannka** [comment]
LGTM. 
Defer to @sriramkrish85 for the final approval.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: So, I'm joining this a bit late and I see (most likely) the original GitHub issue in #3019. Could you please catch me up:

1. Is #3019 indeed the origin issue?
2. Why did we decide to allow `amp-ad` to communicate with child frames directly vs proxying these messages with 3p iframe (most immediate child of `amp-ad`)? I'm not saying that this is wrong, I'd just like to get into right context on this.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 1 above)
Comment (sriramkrish85) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: This should be fixed now. Please rebase.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 2 above)
Comment (nekodo) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Indeed, #3019 is the issue that this PR pertains to.

Direct communication achieves a few different things:

1. It offers better performance, especially for things like IntersectionObserver emulation, as it avoids adding forwarding. It also makes it easier to handle double-nesting without adding extra complexity.
2. It supports A4A use case where the returned ad is a non-AMP creative. In that case the ad is rendered in a 3P iframe, but it does not have access to the normal window.context. With the current implementation of messaging there will be no extra work required to support IntersectionObserver or resizing in that case.

Note that the nested iframes can only communicate if they know a per-3P-frame unique sentinel value. This sentinel is available within the frame's contextWindow and has to be explicitly passed to any child iframes (i.e. it functions as an authorization token).

FYI, the underlying changes to the messaging code happened some time ago in #3327 along with the changes for IntersectionObserver and for page visibility in #4021.

Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: @nekodo Thanks. This all sounds good.

@sriramkrish85 I'm ok with these changes.

**Comment by sriramkrish85** [comment]
@nekodo let me know if you want me to merge this.

**Comment by nekodo** [comment]
@sriramkrish85 Yes please:)

**Pull request merged**